sklearn/ensemble/_hist_gradient_boosting/splitting.pyx

```diff
@@ -388,11 +394,12 @@ cdef class Splitter:
                     &left_indices_buffer[offset_in_buffers[thread_idx]],
                     sizeof(unsigned int) * left_counts[thread_idx]
                 )
-                memcpy(
-                    &sample_indices[right_offset[thread_idx]],
-                    &right_indices_buffer[offset_in_buffers[thread_idx]],
-                    sizeof(unsigned int) * right_counts[thread_idx]
-                )
+                if right_counts[thread_idx] > 0:
+                    memcpy(
+                        &sample_indices[right_offset[thread_idx]],
+                        &right_indices_buffer[offset_in_buffers[thread_idx]],
+                        sizeof(unsigned int) * right_counts[thread_idx]
+                    )
```

Comment:
This was crucial for fixing the out of bound error, but I don't know why this is necessary only with adding this PR while it was not necessary before.

Reply:
Are you referencing the issue with `test_grower_interaction_constraints` mentioned in https://github.com/scikit-learn/scikit-learn/pull/21020#issuecomment-921158509 ? 

(I am having trouble reproducing the out of bound error removing `if` condition.)

Reply:
I can reproduce the `IndexError` at running time (when running the given test as reported in https://github.com/scikit-learn/scikit-learn/pull/21020#issuecomment-922337842) after having compiled with `boundscheck=True`. I haven't been able to identify the cause of this error.

Fortunately, compiling without bounds checks on memory-views (`boundscheck=False`) works and all HGBT tests pass even without this `if`.

Reply:
@jjerphan Wanna have a look at #21130 where this is fixed and described properly?

Reply:
I will have a look at it this afternoon. :+1: